python/cuml/tests/test_device_selection.py

```diff
@@ -0,0 +1,183 @@
+    sk_model.fit(X_train, y_train)
+
+    input_type = request.param['input_type']
+
+    if input_type == 'dataframe':
+        modified_y_train = pd.Series(y_train)
+    elif input_type == 'cudf':
+        modified_y_train = cudf.Series(y_train)
+    else:
+        modified_y_train = to_output_type(y_train, input_type)
+
+    return {
+        'fit_intercept': request.param['fit_intercept'],
+        'normalize': request.param['normalize'],
+        'X_train': to_output_type(X_train, input_type),
+        'y_train': modified_y_train,
+        'X_test': to_output_type(X_test, input_type),
+        'ref_y_test': sk_model.predict(X_test)
+    }
+
+
+def test_train_cpu_infer_cpu(lr_data):
+    model = LinearRegression(fit_intercept=lr_data['fit_intercept'],
+                             normalize=lr_data['normalize'])
+    with using_device_type('cpu'):
+        model.fit(lr_data['X_train'], lr_data['y_train'])
+        cu_pred = model.predict(lr_data['X_test'])
+
+    cu_pred = to_output_type(cu_pred, 'numpy').flatten()
+    np.testing.assert_allclose(lr_data['ref_y_test'], cu_pred)
+
+
+def test_train_gpu_infer_cpu(lr_data):
+    model = LinearRegression(fit_intercept=lr_data['fit_intercept'],
+                             normalize=lr_data['normalize'])
+    with using_device_type('gpu'):
+        model.fit(lr_data['X_train'], lr_data['y_train'])
+    with using_device_type('cpu'):
+        cu_pred = model.predict(lr_data['X_test'])
+
+    cu_pred = to_output_type(cu_pred, 'numpy').flatten()
+    np.testing.assert_allclose(lr_data['ref_y_test'], cu_pred)
+
+
+def test_train_cpu_infer_gpu(lr_data):
+    model = LinearRegression(fit_intercept=lr_data['fit_intercept'],
+                             normalize=lr_data['normalize'])
+    with using_device_type('cpu'):
+        model.fit(lr_data['X_train'], lr_data['y_train'])
+    with using_device_type('gpu'):
+        cu_pred = model.predict(lr_data['X_test'])
+
+    cu_pred = to_output_type(cu_pred, 'numpy').flatten()
+    np.testing.assert_allclose(lr_data['ref_y_test'], cu_pred)
+
+
+def test_train_gpu_infer_gpu(lr_data):
+    model = LinearRegression(fit_intercept=lr_data['fit_intercept'],
+                             normalize=lr_data['normalize'])
+    with using_device_type('gpu'):
+        model.fit(lr_data['X_train'], lr_data['y_train'])
+        cu_pred = model.predict(lr_data['X_test'])
+
+    cu_pred = to_output_type(cu_pred, 'numpy').flatten()
+    np.testing.assert_allclose(lr_data['ref_y_test'], cu_pred)
+    sk_model = skLinearRegression(fit_intercept=lr_data['fit_intercept'],
+                                  normalize=lr_data['normalize'])
+    sk_model.fit(X_train, y_train)
+    sk_pred = sk_model.predict(X_test)
+    np.testing.assert_allclose(sk_pred, cu_pred)
+
+
+@pytest.mark.parametrize('fit_intercept', [False, True])
+@pytest.mark.parametrize('normalize', [False, True])
+def test_pickle_interop(fit_intercept, normalize):
```

Comment:
This seems to be testing pickling and unpickling in the same "device type", which works when this is a modality of the GPU package. i.e. it is pickling using device type `gpu` and then just unpickling, and then doing inference with cpu, which uses the existing capabilities. But we will need to add logic to unpickle GPU objects in the CPU only package, and add tests for that later, no?

Reply:
When we are able to release an exclusive CPU package, we will definitely need to add more tests and robustness. But since we are still undecided on how something like that will look and if that will cause our current interop approach to change, it's definitely open-ended for now